File: examples/pokemon-collection/android/features/collection/src/main/java/com/example/pokemoncollection/collection/CollectionStore.kt
package com.example.pokemoncollection.collection

import android.content.res.AssetManager
import android.icu.text.BreakIterator
import java.util.Locale
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import java.io.File
import java.io.FileOutputStream
import java.nio.ByteBuffer
import java.nio.charset.CodingErrorAction
import java.nio.file.Files
import java.nio.file.NoSuchFileException
import java.nio.file.StandardCopyOption
import java.util.UUID
import org.json.JSONArray
import org.json.JSONObject
import org.json.JSONTokener

public data class Card(val id: String, val name: String, val number: String, val set: String, val printedTotal: Int, val rarity: String, val image: String, val finishes: List<String>) {
    val identity: String get() = "$set · $number/$printedTotal"
}
public data class CopyEntry(val id: String, val cardId: String, val quantity: Int, val finish: String, val condition: String, val note: String, val createdAt: Long)
public data class CollectionState(val copies: List<CopyEntry> = emptyList(), val wishlist: Map<String, Long> = emptyMap()) {
    val totalCopies: Int get() = copies.sumOf { it.quantity }
    val uniqueCards: Int get() = copies.map { it.cardId }.toSet().size
    fun quantity(cardId: String): Int = copies.filter { it.cardId == cardId }.sumOf { it.quantity }
}
public enum class Scope { Collection, Wishlist, Catalog }
public enum class SortOrder(val label: String) { Recent("Recently saved"), Name("Name"), Number("Card number") }
public val copyConditions: List<String> = listOf("Near mint", "Lightly played", "Moderately played", "Heavily played", "Damaged")

public class Catalog(val cards: List<Card>) {
    init { require(cards.isNotEmpty() && cards.map { it.id }.toSet().size == cards.size) }
    fun search(query: String): List<Card> {
        val term = query.trim()
        return cards.filter {
            term.isEmpty() || it.name.contains(term, true) || it.set.contains(term, true) ||
                "Scarlet & Violet–${it.set}".contains(term, true) || it.number.contains(term, true) ||
                (term.toIntOrNull() != null && term.toIntOrNull() == it.number.toIntOrNull())
        }
    }
    companion object {
        fun read(assets: AssetManager): Catalog {
            val array = JSONArray(assets.open("catalog.json").bufferedReader().use { it.readText() })
            return Catalog((0 until array.length()).map { index ->
                val item = array.getJSONObject(index)
                val finishes = item.getJSONArray("finishes")
                Card(item.getString("id"), item.getString("name"), item.getString("number"), item.getString("set"), item.getInt("printedTotal"), item.getString("rarity"), item.getString("image"), (0 until finishes.length()).map { finishes.getString(it) })
            })
        }
    }
}

/** Missing data is distinct from unreadable data. Writes either replace the file atomically or throw. */
public interface CollectionStorage { fun read(): String?; fun write(value: String) }
public class AtomicCollectionStorage(private val file: File) : CollectionStorage {
    override fun read(): String? = try {
        Charsets.UTF_8.newDecoder().onMalformedInput(CodingErrorAction.REPORT).onUnmappableCharacter(CodingErrorAction.REPORT)
            .decode(ByteBuffer.wrap(Files.readAllBytes(file.toPath()))).toString()
    } catch (_: NoSuchFileException) { null }

    override fun write(value: String) {
        Files.createDirectories(requireNotNull(file.parentFile).toPath())
        val temporary = Files.createTempFile(file.parentFile!!.toPath(), ".collection-", ".tmp")
        try {
            FileOutputStream(temporary.toFile()).use { output ->
                output.write(value.toByteArray(Charsets.UTF_8))
                output.fd.sync()
            }
            // No non-atomic fallback: a filesystem without atomic replacement exposes a retryable error.
            Files.move(temporary, file.toPath(), StandardCopyOption.ATOMIC_MOVE, StandardCopyOption.REPLACE_EXISTING)
        } finally { Files.deleteIfExists(temporary) }
    }
}

/** Feature-owned committed state. The app supplies storage; UI dispatches blocking work on IO. */
public class CollectionStore(val catalog: Catalog, private val storage: CollectionStorage) {
    var state: CollectionState by mutableStateOf(CollectionState())
        private set
    var loaded: Boolean by mutableStateOf(false)
        private set
    var error: String? by mutableStateOf(null)
        private set
    var notice: String? by mutableStateOf(null)
        private set
    private var pending: Pair<CollectionState, String>? = null

    @Synchronized fun load(): Boolean = try {
        val restored = storage.read()?.let(::decode) ?: CollectionState()
        state = restored
        loaded = true
        error = null
        pending = null
        true
    } catch (_: Exception) {
        error = "Saved data could not be read. Your data has not been changed. Retry when storage is available."
        notice = null
        false
    }

    @Synchronized fun retry(): Boolean {
        if (!loaded) return load()
        return pending?.let { commit(it.first, it.second) } ?: true
    }

    @Synchronized fun clearFeedback() { if (loaded) error = null; notice = null; pending = null }

    @Synchronized fun addCopy(id: String, quantity: Int, finish: String, condition: String, note: String): Boolean {
        if (!loaded) return false
        val entry = CopyEntry(UUID.randomUUID().toString(), id, quantity, finish, condition, note, System.currentTimeMillis())
        try { validateEntry(entry) } catch (failure: IllegalArgumentException) {
            error = failure.message
            notice = null
            pending = null
            return false
        }
        return commit(state.copy(copies = state.copies + entry), "Copy saved")
    }

    @Synchronized fun setWishlist(id: String, wanted: Boolean): Boolean {
        if (!loaded || catalog.cards.none { it.id == id }) return false
        if (state.wishlist.containsKey(id) == wanted) return true
        val wishlist = if (wanted) state.wishlist + (id to System.currentTimeMillis()) else state.wishlist - id
        return commit(state.copy(wishlist = wishlist), if (wanted) "Added to wishlist" else "Removed from wishlist")
    }

    @Synchronized fun deleteCopy(id: String): Boolean {
        if (!loaded || state.copies.none { it.id == id }) return false
        return commit(state.copy(copies = state.copies.filterNot { it.id == id }), "Copy entry deleted")
    }

    fun cards(scope: Scope, query: String, sort: SortOrder): List<Card> {
        val matches = catalog.search(query).filter { card -> when (scope) {
            Scope.Catalog -> true
            Scope.Collection -> state.quantity(card.id) > 0
            Scope.Wishlist -> state.wishlist.containsKey(card.id)
        } }
        return when (sort) {
            SortOrder.Name -> matches.sortedBy { it.name.lowercase() }
            SortOrder.Number -> matches.sortedBy { it.number.toInt() }
            SortOrder.Recent -> if (scope == Scope.Catalog) matches else matches.sortedByDescending { card ->
                if (scope == Scope.Wishlist) state.wishlist[card.id] ?: 0L
                else state.copies.filter { it.cardId == card.id }.maxOfOrNull { it.createdAt } ?: 0L
            }
        }
    }

    private fun commit(candidate: CollectionState, message: String): Boolean {
        notice = null
        return try {
            storage.write(encode(candidate))
            state = candidate
            pending = null
            error = null
            notice = message
            true
        } catch (_: Exception) {
            pending = candidate to message
            error = "Could not save changes. Your last saved collection is unchanged. Retry to save this change."
            false
        }
    }

    private fun validateEntry(entry: CopyEntry) {
        val card = catalog.cards.find { it.id == entry.cardId }
        require(card != null) { "Choose a card from the example catalog." }
        require(entry.quantity in 1..99) { "Quantity must be between 1 and 99." }
        require(entry.finish in card.finishes) { "Choose an available finish." }
        require(entry.condition in copyConditions) { "Choose a condition." }
        require(noteCharacterCount(entry.note) <= 500) { "Keep the note to 500 characters or fewer." }
        require(entry.createdAt > 0) { "Invalid saved date." }
        UUID.fromString(entry.id)
    }

    private fun encode(value: CollectionState): String = JSONObject().apply {
        put("version", 1)
        put("copies", JSONArray().apply { value.copies.forEach { entry -> put(JSONObject().apply {
            put("id", entry.id); put("cardId", entry.cardId); put("quantity", entry.quantity)
            put("finish", entry.finish); put("condition", entry.condition); put("note", entry.note); put("createdAt", entry.createdAt)
        }) } })
        put("wishlist", JSONArray().apply { value.wishlist.forEach { (cardId, savedAt) -> put(JSONObject().apply {
            put("cardId", cardId); put("savedAt", savedAt)
        }) } })
    }.toString()

    private fun decode(raw: String): CollectionState {
        val tokens = JSONTokener(raw)
        val root = tokens.nextValue() as? JSONObject ?: error("Invalid saved data")
        require(tokens.nextClean() == '\u0000') { "Trailing saved data" }
        require(root.strictLong("version") == 1L) { "Unsupported saved data version." }
        val copyArray = root.getJSONArray("copies")
        val copies = (0 until copyArray.length()).map { index ->
            val item = copyArray.getJSONObject(index)
            val quantity = item.strictLong("quantity")
            require(quantity in 1..99)
            CopyEntry(item.strictString("id"), item.strictString("cardId"), quantity.toInt(), item.strictString("finish"), item.strictString("condition"), item.strictString("note"), item.strictLong("createdAt")).also(::validateEntry)
        }
        require(copies.map { it.id }.toSet().size == copies.size)
        val wishArray = root.getJSONArray("wishlist")
        val wishes = (0 until wishArray.length()).map { index ->
            val item = wishArray.getJSONObject(index)
            val id = item.strictString("cardId")
            val time = item.strictLong("savedAt")
            require(catalog.cards.any { it.id == id } && time > 0)
            id to time
        }
        require(wishes.map { it.first }.toSet().size == wishes.size)
        require(copies.sumOf { it.quantity.toLong() } <= Int.MAX_VALUE)
        return CollectionState(copies, wishes.toMap())
    }
}
private fun JSONObject.strictString(key: String): String = (get(key) as? String) ?: error("Invalid string")
private fun JSONObject.strictLong(key: String): Long {
    val value = get(key)
    require(value is Int || value is Long) { "Invalid integer" }
    return (value as Number).toLong()
}

/** ICU grapheme boundaries match user-visible characters, including emoji sequences and accents. */
internal fun noteCharacterCount(note: String): Int {
    val boundaries = BreakIterator.getCharacterInstance(Locale.ROOT)
    boundaries.setText(note)
    boundaries.first()
    var count = 0
    while (boundaries.next() != BreakIterator.DONE) count++
    return count
}

File: examples/pokemon-collection/ios/Packages/CollectionFeature/Sources/CollectionFeature/CollectionStore.swift
import Foundation
import Observation

public struct Card: Codable, Identifiable, Hashable, Sendable {
    public let id: String
    public let name: String
    public let number: String
    public let set: String
    public let printedTotal: Int
    public let rarity: String
    public let image: String
    public let finishes: [String]
}
public enum CardCondition: String, Codable, CaseIterable, Sendable {
    case nearMint = "Near mint", lightlyPlayed = "Lightly played", moderatelyPlayed = "Moderately played", heavilyPlayed = "Heavily played", damaged = "Damaged"
}
public struct CopyEntry: Codable, Identifiable, Equatable, Sendable {
    public let id: UUID
    public let cardID: String
    public let quantity: Int
    public let finish: String
    public let condition: CardCondition
    public let note: String
    public let addedAt: Date
}
public struct WishlistEntry: Codable, Equatable, Sendable {
    public let cardID: String
    public let addedAt: Date
}
public struct CollectionSnapshot: Codable, Equatable, Sendable {
    var schemaVersion = 1
    public var copies: [CopyEntry] = []
    public var wishlist: [WishlistEntry] = []
}
public enum CardScope: Sendable { case collection, wishlist, catalog }
public enum CardSort: String, CaseIterable, Sendable { case recent = "Recently added", name = "Name", number = "Card number" }
@MainActor public protocol CollectionStorage {
    func read() throws -> Data?
    func write(_ data: Data) throws
}
/// Synchronous, main-actor storage serializes this small local document with UI mutations.
/// The replacement is atomic: a failed write cannot replace the committed document.
@MainActor public final class FileCollectionStorage: CollectionStorage {
    private let url: URL
    public init(url: URL) { self.url = url }
    public func read() throws -> Data? {
        do { return try Data(contentsOf: url) }
        catch let error as CocoaError where error.code == .fileReadNoSuchFile { return nil }
    }
    public func write(_ data: Data) throws {
        try FileManager.default.createDirectory(at: url.deletingLastPathComponent(), withIntermediateDirectories: true)
        try data.write(to: url, options: .atomic)
    }
}

public struct CollectionError: LocalizedError, Sendable {
    public let message: String
    public var errorDescription: String? { message }
    init(_ message: String) { self.message = message }
}

public enum CardCatalog {
    public static func load(from url: URL) throws -> [Card] {
        let cards = try JSONDecoder().decode([Card].self, from: Data(contentsOf: url))
        guard !cards.isEmpty, Set(cards.map(\.id)).count == cards.count,
              cards.allSatisfy({ !$0.id.isEmpty && !$0.name.isEmpty && Int($0.number) != nil && !$0.set.isEmpty && $0.printedTotal > 0 && !$0.finishes.isEmpty && Set($0.finishes).count == $0.finishes.count && Set($0.finishes).isSubset(of: ["Normal", "Reverse holo", "Holo"]) && URL(fileURLWithPath: $0.image).lastPathComponent == $0.image }) else {
            throw CollectionError("The bundled card catalog is invalid.")
        }
        return cards
    }
}

@Observable @MainActor public final class CollectionStore {
    public let catalog: [Card]
    public private(set) var snapshot = CollectionSnapshot()
    public private(set) var loadError: String?
    private let storage: any CollectionStorage

    public init(catalog: [Card], storage: any CollectionStorage) {
        self.catalog = catalog
        self.storage = storage
        reload()
    }
    public var totalCopies: Int { snapshot.copies.reduce(0) { $0 + $1.quantity } }
    public var uniqueCards: Int { Set(snapshot.copies.map(\.cardID)).count }
    public func copies(of id: String) -> [CopyEntry] { snapshot.copies.filter { $0.cardID == id } }
    public func quantity(of id: String) -> Int { copies(of: id).reduce(0) { $0 + $1.quantity } }
    public func isWishlisted(_ id: String) -> Bool { snapshot.wishlist.contains { $0.cardID == id } }

    public func reload() {
        do {
            let loaded = try storage.read().map { try JSONDecoder().decode(CollectionSnapshot.self, from: $0) } ?? CollectionSnapshot()
            try validate(loaded)
            snapshot = loaded
            loadError = nil
        } catch {
            loadError = "Your saved collection could not be read. The existing data has been preserved. \(error.localizedDescription)"
        }
    }

    public func cards(matching query: String = "", scope: CardScope, sort: CardSort = .recent) -> [Card] {
        let term = query.trimmingCharacters(in: .whitespacesAndNewlines)
        let matches = catalog.filter { card in
            let inScope = scope == .catalog || (scope == .collection ? quantity(of: card.id) > 0 : isWishlisted(card.id))
            let numberMatches = Int(term).map { $0 == Int(card.number) } ?? card.number.localizedCaseInsensitiveContains(term)
            return inScope && (term.isEmpty || card.name.localizedCaseInsensitiveContains(term) || card.set.localizedCaseInsensitiveContains(term) || "Scarlet & Violet—\(card.set)".localizedCaseInsensitiveContains(term) || numberMatches)
        }
        if scope == .catalog && sort == .recent { return matches }
        return matches.sorted { left, right in
            switch sort {
            case .name: return left.name.localizedStandardCompare(right.name) == .orderedAscending
            case .number: return (Int(left.number) ?? 0) < (Int(right.number) ?? 0)
            case .recent:
                let a = latestDate(for: left.id, scope: scope)
                let b = latestDate(for: right.id, scope: scope)
                return a == b ? left.id < right.id : a > b
            }
        }
    }

    public func add(cardID: String, quantity: Int, finish: String, condition: CardCondition, note: String) throws {
        let entry = CopyEntry(id: UUID(), cardID: cardID, quantity: quantity, finish: finish, condition: condition, note: note, addedAt: Date())
        try validate(entry)
        var next = snapshot
        next.copies.append(entry)
        try commit(next)
    }

    /// Setting membership is idempotent and never changes owned copies.
    public func setWishlist(cardID: String, included: Bool) throws {
        try requireReadable()
        guard catalog.contains(where: { $0.id == cardID }) else { throw CollectionError("Unknown card.") }
        guard isWishlisted(cardID) != included else { return }
        var next = snapshot
        next.wishlist.removeAll { $0.cardID == cardID }
        if included { next.wishlist.append(WishlistEntry(cardID: cardID, addedAt: Date())) }
        try commit(next)
    }

    public func delete(entryID: UUID) throws {
        var next = snapshot
        guard next.copies.contains(where: { $0.id == entryID }) else { throw CollectionError("This copy entry no longer exists.") }
        next.copies.removeAll { $0.id == entryID }
        try commit(next)
    }

    private func commit(_ next: CollectionSnapshot) throws {
        try requireReadable()
        try validate(next)
        do { try storage.write(JSONEncoder().encode(next)) }
        catch { throw CollectionError("Could not save your collection. Your previous saved collection is unchanged. Try again. \(error.localizedDescription)") }
        snapshot = next
    }
    private func requireReadable() throws {
        guard loadError == nil else { throw CollectionError("Retry loading your collection before making changes.") }
    }
    private func validate(_ entry: CopyEntry) throws {
        guard let card = catalog.first(where: { $0.id == entry.cardID }) else { throw CollectionError("Unknown card in saved collection.") }
        guard (1...99).contains(entry.quantity) else { throw CollectionError("Quantity must be from 1 to 99.") }
        guard card.finishes.contains(entry.finish) else { throw CollectionError("Choose a finish available for this card.") }
        guard entry.note.count <= 500 else { throw CollectionError("Notes must be 500 characters or fewer.") }
        guard entry.addedAt.timeIntervalSinceReferenceDate.isFinite else { throw CollectionError("Invalid saved date.") }
    }
    private func validate(_ document: CollectionSnapshot) throws {
        guard document.schemaVersion == 1 else { throw CollectionError("This collection uses an unsupported storage version.") }
        guard Set(document.copies.map(\.id)).count == document.copies.count,
              Set(document.wishlist.map(\.cardID)).count == document.wishlist.count else { throw CollectionError("The saved collection has duplicate entries.") }
        for entry in document.copies { try validate(entry) }
        for entry in document.wishlist {
            guard catalog.contains(where: { $0.id == entry.cardID }), entry.addedAt.timeIntervalSinceReferenceDate.isFinite else { throw CollectionError("Invalid saved wishlist entry.") }
        }
    }
    private func latestDate(for id: String, scope: CardScope) -> Date {
        if scope == .wishlist { return snapshot.wishlist.first { $0.cardID == id }?.addedAt ?? .distantPast }
        return copies(of: id).map(\.addedAt).max() ?? .distantPast
    }
}
